packages/firestore/src/util/async_queue.ts

```diff
@@ -178,25 +206,35 @@ export class AsyncQueue {
   }
 
   /**
-   * Schedules an operation to be run on the AsyncQueue once the specified
-   * `delayMs` has elapsed. The returned DelayedOperationResult can be
-   * used to cancel the operation prior to its running.
+   * Schedules an operation to be queued on the AsyncQueue once the specified
+   * `delayMs` has elapsed. The returned CancelablePromise can be used to cancel
+   * the operation prior to its running.
    */
-  scheduleWithDelay<T>(
-    op: () => Promise<T>,
-    delayMs: number
+  enqueueAfterDelay<T>(
+    timerId: TimerId,
+    delayMs: number,
+    op: () => Promise<T>
   ): CancelablePromise<T> {
     this.verifyNotFailed();
 
-    const delayedOp = DelayedOperation.createAndSchedule(this, op, delayMs);
+    // While not necessarily harmful, we currently don't expect to have multiple
+    // ops with the same timer id in the queue, so defensively reject them.
+    assert(
+      !this.hasDelayedOperation(timerId),
+      `Attempted to schedule multiple operations with timer id ${
+        TimerId[timerId]
+      }.`
+    );
+
+    const delayedOp = DelayedOperation.createAndSchedule(
+      this,
+      timerId,
+      delayMs,
+      op,
+      op => this.removeDelayedOperation(op)
+    );
     this.delayedOperations.push(delayedOp);
 
-    delayedOp.catch(err => {}).then(() => {
```

Comment:
I think this way of dealing with the dequeuing was much more straightforward (albeit you need to re-throw the error)